File: src/main/java/toby/command/commands/fetch/MemeCommand.java
package toby.command.commands.fetch;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.interactions.commands.OptionMapping;
import net.dv8tion.jda.api.interactions.commands.OptionType;
import net.dv8tion.jda.api.interactions.commands.build.OptionData;
import org.apache.http.HttpResponse;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.HttpClients;
import org.jetbrains.annotations.NotNull;
import toby.command.CommandContext;
import toby.dto.web.RedditAPIDto;
import toby.jpa.dto.UserDto;

import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.Random;

import static toby.command.ICommand.deleteAfter;
import static toby.command.ICommand.getConsumer;

public class MemeCommand implements IFetchCommand {

    private final String SUBREDDIT = "subreddit";
    private final String TIME_PERIOD = "timeperiod";
    private final String LIMIT = "limit";

    @Override
    public void handle(CommandContext ctx, UserDto requestingUserDto, Integer deleteDelay) {
        try {
            handle(ctx, HttpClients.createDefault(), requestingUserDto, deleteDelay);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public void handle(CommandContext ctx, HttpClient httpClient, UserDto requestingUserDto, Integer deleteDelay) throws IOException {
        SlashCommandInteractionEvent event = ctx.getEvent();
        event.deferReply().queue();
        deleteAfter(event.getHook(), deleteDelay);
        if (!requestingUserDto.hasMemePermission()) {
            sendErrorMessage(event, deleteDelay);
            return;
        }
        RedditApiArgs result = getRedditApiArgs(event);
        if (result.subredditArgOptional().isPresent()) {
            String subredditArg = result.subredditArgOptional().get();
            if (subredditArg.equals("sneakybackgroundfeet")) {
                event.getChannel().sendMessageFormat("Don't talk to me.").queue(getConsumer(deleteDelay));
            } else {
                MessageEmbed embed = fetchRedditPost(result, event, deleteDelay, httpClient);
                event.getChannel().sendMessageEmbeds(embed).queue();
            }
        }
    }

    @NotNull
    private RedditApiArgs getRedditApiArgs(SlashCommandInteractionEvent event) {
        Optional<String> subredditArgOptional = Optional.ofNullable(event.getOption(SUBREDDIT)).map(OptionMapping::getAsString);
        Optional<String> timePeriodOptional = Optional.ofNullable(event.getOption(TIME_PERIOD)).map(OptionMapping::getAsString);
        String timePeriod = RedditAPIDto.TimePeriod.valueOf(timePeriodOptional.orElse("DAY")).toString().toLowerCase();
        int limit = Optional.ofNullable(event.getOption(LIMIT)).map(OptionMapping::getAsInt).orElse(5);

        return new RedditApiArgs(subredditArgOptional, timePeriod, limit);
    }

    private MessageEmbed fetchRedditPost(RedditApiArgs result, SlashCommandInteractionEvent event, int deleteDelay, HttpClient httpClient) throws IOException {
        Gson gson = new Gson();
        String redditApiUrl = String.format(RedditAPIDto.redditPrefix, result.subredditArgOptional.get(), result.limit(), result.timePeriod());
        HttpGet request = new HttpGet(redditApiUrl);
        HttpResponse response = httpClient.execute(request);

        // Check the response code (200 indicates success)
        if (response.getStatusLine().getStatusCode() == 200) {
            // Parse the JSON response
            JsonObject jsonResponse = JsonParser.parseReader(new InputStreamReader(response.getEntity().getContent())).getAsJsonObject();

            JsonArray children = jsonResponse.getAsJsonObject("data").getAsJsonArray("children");
            Random random = new Random();

            // Extract the data you need from the JSON
            JsonObject meme = children
                    .get(random.nextInt(children.size()))
                    .getAsJsonObject()
                    .getAsJsonObject("data");
            RedditAPIDto redditAPIDto = gson.fromJson(meme.toString(), RedditAPIDto.class);
            if (redditAPIDto.isNsfw()) {
                event.getHook().sendMessageFormat("I received a NSFW subreddit from %s, or reddit gave me a NSFW meme, either way somebody shoot that guy", event.getMember()).queue(getConsumer(deleteDelay));
            } else if (redditAPIDto.getVideo()) {
                event.getHook().sendMessageFormat("I pulled back a video, whoops. Try again maybe? Or not, up to you.").queue(getConsumer(deleteDelay));
            }

            String title = meme.get("title").getAsString();
            String url = meme.get("url").getAsString();
            String author = meme.get("author").getAsString();

            // Create a JDA EmbedBuilder and return the MessageEmbed
            EmbedBuilder embedBuilder = new EmbedBuilder();
            embedBuilder.setTitle(title, url);
            embedBuilder.setAuthor(author, null, null);
            embedBuilder.setImage(url);
            embedBuilder.addField("subreddit", result.subredditArgOptional.get(), false);

            return embedBuilder.build();
        }

        return null; // Return null if there was an error

    }

    private record RedditApiArgs(Optional<String> subredditArgOptional, String timePeriod, int limit) {
    }


    @Override
    public String getName() {
        return "meme";
    }

    @Override
    public String getDescription() {
        return "This command shows a meme from the subreddit you've specified (SFW only)";
    }

    @Override
    public List<OptionData> getOptionData() {
        OptionData subreddit = new OptionData(OptionType.STRING, SUBREDDIT, "Which subreddit to pull the meme from", true);
        OptionData timePeriod = new OptionData(OptionType.STRING, TIME_PERIOD, "What time period filter to apply to the subreddit (e.g. day/week/month/all). Default day.", false);
        Arrays.stream(RedditAPIDto.TimePeriod.values()).forEach(tp -> timePeriod.addChoice(tp.getTimePeriod(), tp.name()));
        OptionData limit = new OptionData(OptionType.INTEGER, LIMIT, "Pick from top X posts of that day. Default 5.", false);

        return List.of(subreddit, timePeriod, limit);
    }
}

File: src/main/kotlin/toby/command/commands/fetch/MemeCommand.kt
package toby.command.commands.fetch

import com.google.gson.Gson
import com.google.gson.JsonParser
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.interactions.commands.OptionType
import net.dv8tion.jda.api.interactions.commands.build.OptionData
import org.apache.http.client.HttpClient
import org.apache.http.client.methods.HttpGet
import org.apache.http.impl.client.HttpClients
import org.apache.http.util.EntityUtils
import toby.command.CommandContext
import toby.command.ICommand.Companion.invokeDeleteOnMessageResponse
import toby.dto.web.RedditAPIDto
import toby.dto.web.RedditAPIDto.TimePeriod
import toby.jpa.dto.UserDto
import java.io.IOException
import java.io.InputStreamReader
import kotlin.random.Random

class MemeCommand : IFetchCommand {
    private val SUBREDDIT = "subreddit"
    private val TIME_PERIOD = "timeperiod"
    private val LIMIT = "limit"

    override fun handle(ctx: CommandContext, requestingUserDto: UserDto, deleteDelay: Int?) {
        try {
            logger.setGuildAndMemberContext(ctx.guild, ctx.member)
            handle(ctx, HttpClients.createDefault(), requestingUserDto, deleteDelay)
        } catch (e: IOException) {
            logger.error("IOException occurred while handling command: $e")
            throw RuntimeException(e)
        }
    }

    @Throws(IOException::class)
    fun handle(ctx: CommandContext, httpClient: HttpClient, requestingUserDto: UserDto?, deleteDelay: Int?) {
        val event = ctx.event
        logger.setGuildAndMemberContext(ctx.guild, ctx.member)
        event.deferReply().queue()

        if (requestingUserDto?.memePermission != true) {
            logger.info("User ${event.member?.effectiveName} does not have meme permission")
            sendErrorMessage(event, deleteDelay ?: 0)
            return
        }

        val result = getRedditApiArgs(event)
        val subredditArg = result.subredditArg

        if (subredditArg == "sneakybackgroundfeet") {
            logger.info("Requested subreddit 'sneakybackgroundfeet'")
            event.hook.sendMessageFormat("Don't talk to me.").queue(invokeDeleteOnMessageResponse(deleteDelay ?: 0))
        } else {
            val embed = fetchRedditPost(result, event, deleteDelay ?: 0, httpClient)
            if (embed != null) {
                logger.info("Successfully fetched meme from subreddit: $subredditArg")
                event.hook.sendMessageEmbeds(embed).queue()
            } else {
                logger.warn("Failed to fetch meme from subreddit: $subredditArg")
                event.hook.sendMessage("Failed to fetch meme. Please try again later.")
                    .queue(invokeDeleteOnMessageResponse(deleteDelay ?: 0))
            }
        }
    }

    private fun getRedditApiArgs(event: SlashCommandInteractionEvent): RedditApiArgs {
        val subredditArg = event.getOption(SUBREDDIT)?.asString
        val timePeriod = TimePeriod.parseTimePeriod(event.getOption(TIME_PERIOD)?.asString ?: "day").timePeriod
        val limit = event.getOption(LIMIT)?.asInt ?: 5
        logger.info { "Reddit API args - Subreddit: $subredditArg, Time Period: $timePeriod, Limit: $limit" }
        return RedditApiArgs(subredditArg, timePeriod, limit)
    }

    @Throws(IOException::class)
    private fun fetchRedditPost(
        result: RedditApiArgs,
        event: SlashCommandInteractionEvent,
        deleteDelay: Int,
        httpClient: HttpClient
    ): MessageEmbed? {
        val gson = Gson()
        val redditApiUrl =
            String.format(RedditAPIDto.redditPrefix, result.subredditArg, result.limit, result.timePeriod)
        val request = HttpGet(redditApiUrl)
        logger.info("Fetching Reddit post from URL: $redditApiUrl")
        val response = httpClient.execute(request)

        return response.let { // Using `also` to perform additional operations on the response object
            if (response.statusLine.statusCode == 200) {
                val jsonResponse = JsonParser.parseReader(InputStreamReader(response.entity.content)).asJsonObject
                EntityUtils.consume(response.entity) // Ensure the entity content is fully consumed

                val children = jsonResponse.getAsJsonObject("data").getAsJsonArray("children")

                if (children.size() == 0) {
                    logger.info("No memes found in subreddit: ${result.subredditArg}")
                    event.hook.sendMessageFormat("No memes found in the subreddit.")
                        .queue(invokeDeleteOnMessageResponse(deleteDelay))
                    return null
                }

                val meme = children[Random.nextInt(children.size())].asJsonObject.getAsJsonObject("data")
                val redditAPIDto = gson.fromJson(meme.toString(), RedditAPIDto::class.java)

                if (redditAPIDto.isNsfw == true) {
                    logger.warn("NSFW meme detected from subreddit: ${result.subredditArg}")
                    event.hook.sendMessageFormat(
                        "I received a NSFW subreddit from %s, or reddit gave me a NSFW meme, either way somebody shoot that guy",
                        event.member
                    ).queue(invokeDeleteOnMessageResponse(deleteDelay))
                    return null
                } else if (redditAPIDto.video == true) {
                    logger.warn("Video meme detected from subreddit: ${result.subredditArg}")
                    event.hook.sendMessageFormat("I pulled back a video, whoops. Try again maybe? Or not, up to you.")
                        .queue(invokeDeleteOnMessageResponse(deleteDelay))
                    return null
                }

                val title = meme["title"].asString
                val url = meme["url"].asString
                val author = meme["author"].asString

                logger.info("Meme fetched successfully - Title: $title, Author: $author, URL: $url")

                return EmbedBuilder().apply {
                    setTitle(title, url)
                    setAuthor(author, null, null)
                    setImage(url)
                    addField("subreddit", result.subredditArg ?: "unknown", false)
                }.build()
            } else {
                logger.error("Error response from Reddit API: ${response.statusLine.statusCode}")
                null
            }
        }
    }

    private data class RedditApiArgs(val subredditArg: String?, val timePeriod: String, val limit: Int)

    override val name: String
        get() = "meme"
    override val description: String
        get() = "This command shows a meme from the subreddit you've specified (SFW only)"
    override val optionData: List<OptionData>
        get() {
            val subreddit = OptionData(OptionType.STRING, SUBREDDIT, "Which subreddit to pull the meme from", true)
            val timePeriod = OptionData(
                OptionType.STRING,
                TIME_PERIOD,
                "What time period filter to apply to the subreddit (e.g. day/week/month/all). Default day.",
                false
            )
            TimePeriod.entries.forEach { tp -> timePeriod.addChoice(tp.timePeriod, tp.timePeriod) }
            val limit = OptionData(OptionType.INTEGER, LIMIT, "Pick from top X posts of that day. Default 5.", false)
            return listOf(subreddit, timePeriod, limit)
        }
}
